torch/distributed/_functional_collectives.py

```diff
@@ -330,6 +330,7 @@ def __new__(cls, elem: torch.Tensor):
         return r
 
     def __repr__(self):
+        wait_tensor(self.elem)
```

Comment:
Hmmm, I can see the rationale behind this change. The reason we have this tensor wrapper is for functional collectives which enable the asynchronous collective call by default. For example, users call all-reduce but does not use it immediately and we want to somehow delay it and get some overlap with computation. So if we call wait in print, then it will break the benefit of delaying of collectives, interfering the original behavior. So I am not quite sure about this change.

Reply:
Do we expect users to include a `print()` in the code when caring about performance?

Reply:
Well, users might put logging in their code? No?

Reply:
If you `print()` a normal tensor, that is a CPU-GPU sync since the print output depends on the values of the tensor. In my understanding, people should not be putting `print()` in the performance-critical part of their code. There may be some logging on scalar tensors at the end of the iteration or on non-scalar tensors every `k` iterations (for some large `k`).

I think that this added `wait_tensor(self.elem)` call makes sense, as it is needed for `print()` correctness.

Reply:
^ this was my reasoning too

Reply:
I see. It makes sense to have the behavior of `print` same as tensor. The original comment is not an objection but rather a question. I just thought there might be some uses cases when users don't want to call wait when calling `print`. But it sounds like that is not the case.